### ISSUE TITLE
udev rule not working

### ISSUE THREAD

**Issue by Germar** [bit-team/backintime#466]
I've set up backintime to work when a drive gets connected. However, it doesn't appear to be working properly. I originally tried with root, since I wish to back up my entire file system. I also tried a minimal situation for my normal user, but this also fails. I'm using KDE Plasma 5.3.1 in Arch Linux.

When I plug in my hard drive, a udev rule appears to activate, because a notification pops up. This says that it cannot find the hard drive, and will wait 30 seconds. Presumably this is because the hard drive is not mounted until I click mount in the Plasma's tray popup. I mount the hard drive within a few seconds, but the backup never starts.

~/.config/backintime/config is here: http://pastebin.com/evrkYs80

/etc/udev/rules.d/99-backintime-sparhawk.rules comprises the following line.

ACTION=="add", ENV{ID_FS_UUID}=="1bf6eef9-f221-4401-a00a-e37b455c8fab", RUN+="/bin/su - 'sparhawk' -c '/usr/bin/nice -n 19 /usr/bin/ionice -c2 -n7 /usr/bin/backintime --backup-job >/dev/null 2>&1'"

There is also /etc/udev/rules.d/99-backintime-root.rules which is almost identical except for the the username changing.

I cannot find any kind of related log file in /var/log

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2015-06-24T23:44:40Z
 * owner: sparhawkthesecond
 * the launchpad url was https://bugs.launchpad.net/bugs/1468553

**Comment by Germar** [comment]
(by germar)
Please activate 'Automount on Attach' for that drive in System Settings > Removeable Devices

**Comment by Germar** [comment]
(by sparhawkthesecond)
I enabled automount; I had to follow these instructions http://askubuntu.com/a/613539/53508

Now it's still not backing up, but I'm not even getting backintime's notifications. I troubleshooted by creating a dummy script echoing the date to a temporary file, using the same triggers as backintime's udev rule. This worked fine, so it seems that the udev rule should be working.

**Comment by Suor** [comment]
Also have this issue. No notifications. Automount is not an option - I use luks encrypted drive with passphrase.

**Comment by Germar** [comment]
Not sure about this, but IMHO the udev rule should fire AFTER opening luks and not as soon as you plug in the drive. I'll have a look

**Comment by Suor** [comment]
Investigated this a bit. There are 2 devices one for container and one for encrypted partition. BackInTime attaches to the latter, which makes sense. 

Unfortunately udev fails. I tried replacing your command in `/etc/udev/rules.d/99-backintime-root.rules` with `touch /home/suor/udev` and that was not fired. I also attached to outer container and it also didn't fire. When I checked that with ordinary unencrypted vfat USB flash drive and it finally worked.

Here is my udev status, comments describe my actions:

```shell
suor:~$ sudo service udev status
● systemd-udevd.service - udev Kernel Device Manager
   Loaded: loaded (/lib/systemd/system/systemd-udevd.service; static; vendor preset: enabled)
   Active: active (running) since Sun 2015-11-08 09:54:05 KRAT; 1min 44s ago
     Docs: man:systemd-udevd.service(8)
           man:udev(7)
 Main PID: 17642 (systemd-udevd)
   CGroup: /system.slice/systemd-udevd.service
           ├─17642 /lib/systemd/systemd-udevd
           ├─18839 /lib/systemd/systemd-udevd
           ├─18840 /lib/systemd/systemd-udevd
           ├─18841 /lib/systemd/systemd-udevd
           └─18842 /lib/systemd/systemd-udevd

Nov 08 09:54:05 suor-xps systemd[1]: Starting udev Kernel Device Manager...
Nov 08 09:54:05 suor-xps systemd-udevd[17642]: starting version 219
Nov 08 09:54:05 suor-xps systemd-udevd[17642]: specified user '<~/>' unknown
Nov 08 09:54:05 suor-xps systemd[1]: Started udev Kernel Device Manager.
# plugging hdd
Nov 08 09:54:28 suor-xps systemd-udevd[17642]: error opening USB device 'descriptors' file
# unmounting hdd (nothing is written here)
# unplugging hdd
Nov 08 09:54:51 suor-xps systemd-udevd[17642]: error opening USB device 'descriptors' file
# plugging flash drive
Nov 08 09:54:59 suor-xps mtp-probe[18315]: checking bus 1, device 10: "/sys/devices/pci0000:00/0000:00:14.0/usb1/1-1"
# unmounting flash drive (nothing is written here)
# unplugging flash drive
Nov 08 09:55:15 suor-xps systemd-udevd[17642]: error opening USB device 'descriptors' file
```

Thing to note here - error is happening when hdd is plugged, but not when unencrypted flash drive is.

Also looked into `/var/log/udev`, nothing relevant there.

**Comment by Suor** [comment]
Found a possible workaround for this - attach by device attributes rather than UUID. This is my udef config:
```shell
ACTION=="add", KERNEL=="sd?1", ATTRS{model}=="BUP Slim SL     ", RUN+="/bin/su - 'root' -c '/usr/bin/nice -n 19 /usr/bin/ionice -c2 -n7 /usr/bin/backintime --backup-job >/home/suor/backintime.log 2>&1'"
```
Things to look at here:
- match is made by device model,
- `KERNEL=="sd?1"` selector is used to single out `/dev/sda1`, in the absense of that event fires for `/dev/sda1`, `/dev/sda`, several usb and pci devices (7 in total for me)
- `backup-job` is replaced with `--backup-job` to trigger detach in `/usr/bin/backintime`

I also added `sleep 10` to `run()` in `/usr/bin/backintime` to wait until volume is mounted. For general use 10 seconds won't be enough (remember about the passphrase prompt), a while loop with check, sleep and eventual give up would be better.

To introspect device attributes this command could be used:
```bash
udevadm info -a -p  $(udevadm info -q path -n /dev/sda1)
```

**Comment by Germar** [comment]
>  backup-job is replaced with --backup-job to trigger detach in /usr/bin/backintime 

That might be the reason. I changed argument parsing with last release but forgot to fix `/usr/bin/backintime` to use both `backup-job` and `--backup-job` for backwards compatibility.

**Comment by LauritzThaulow** [comment]
I'm using Back in Time 1.0.36 on Ubuntu 15.10, and had a very similar issue. I also use a LUKS encrypted USB drive.

I tried to connect the drive while running `udevadm monitor -k -u -p`, and noticed that no `add` event with the specified ID_FS_UUID actually occurred. But I *did* see a `change` event for the drive.

When I changed `add` to `change` in the rule, I saw this in my syslog, and the previous runs had no such messages:

    Jan 15 20:06:42 anacron[26100]: Job `1_Main_profile' terminated
    Jan 15 20:06:42 anacron[26100]: Normal exit (1 job run)

So that change seemed to fix it, in the LUKS case. I just performed a backup so my guess is that it exited quickly without doing anything since there was no need to make more backups today. I will try again tomorrow and see how it goes.
